### DISCUSSION TITLE
async deference operator

### DISCUSSION THREAD

**Discussion by benaadams** [dotnet/csharplang#1117] [General]
async deference operator
----

### Current
await on results:
```csharp
var order = await context.GetOrderAsync(id: 5);
var customer = await order.CustomerAsync();
var addresses = await customer.AddressesAsync();
// + ConfigureAwait
var order = await context.GetOrderAsync(id: 5).ConfigureAwait(false);
var customer = await order.CustomerAsync().ConfigureAwait(false);
var addresses = await customer.AddressesAsync().ConfigureAwait(false);
var address = addresses.FirstOrDefault();
```

Or
```csharp
var address = await (await (await context.GetOrderAsync(id: 5))
                .CustomerAsync()).AddressesAsync())
                .FirstOrDefault();
// + ConfigureAwait
var address = await ( await ( await 
                    context.GetOrderAsync(id: 5).ConfigureAwait(false))
                    .CustomerAsync().ConfigureAwait(false))
                    .AddressesAsync().ConfigureAwait(false)
                    .FirstOrDefault();
```

### Suggested
`->` => `await` + `.ConfigureAwait(false)`
```csharp
var addresses = context.GetOrderAsync(id: 5)
                      ->CustomerAsync()
                      ->AddressesAsync()
                      ->FirstOrDefault();
```

`@.` => `await`
```csharp
var addresses = context.GetOrderAsync(id: 5)
                      @.CustomerAsync()
                      @.AddressesAsync()
                      @.FirstOrDefault();
```

### Considerations

nullables?

`->?`, `@?.`

**Comment by jnm2** [comment]
Related: chaining syntaxes in https://github.com/dotnet/csharplang/issues/827

**Comment by Rattenkrieg** [comment]
I really missing uniform syntax for monad-like computations like Haskell, Scala, F# and others have. 😿

**Comment by Rattenkrieg** [comment]
Abusing LINQ doesn't count.

**Comment by Joe4evr** [comment]
👍 for some syntactic sugar to make chaining tasks nicer.
😕 at the suggested syntax. I'll defer (😉) that decision to the language team.

**Comment by Thaina** [comment]
Will `->` could have problem with pointer?

**Comment by Joe4evr** [comment]
@Thaina That's part of why I'm unsure about the suggested syntax, but the language team can decide whether or not it could create issues.

**Comment by mharasimowicz** [comment]
Related: #74 



```csharp

var addresses = await context.GetOrderAsync(id: 5)
                      |> CustomerAsync()
                      |> AddressesAsync()
                      |> FirstOrDefault();
```

or simpler (depending on chosen syntax):

```csharp

var addresses = await context.GetOrderAsync(id: 5)
                      |> CustomerAsync
                      |> AddressesAsync
                      |> FirstOrDefault;
```

**Comment by benaadams** [comment]
@Thaina 
Contextually no, since `->` for [pointer deference](https://docs.microsoft.com/en-us/dotnet/csharp/language-reference/operators/dereference-operator) can only be used in an `unsafe` context and you can't `await` in an unsafe context; so there would be no context overlap.

There may be an issue with its use changing in different contexts

However, conceptually they are similar

> The `->` operator combines pointer dereferencing and member access.

So `(*x).y` becomes `x->y`

Likewise I'm suggesting it combines task-like dereferencing/awaiting (obtaining its value) and member access.

`(await x.ConfigureAwait(false)).y` becomes `x->y`

**Comment by benaadams** [comment]
@mharasimowicz 
`|>` is more of a pipe operator? Move result into next function; rather than get current value?

**Comment by sharwell** [comment]
The main thing that makes me uncomfortable with this proposal is the use of the `ConfigureAwait(false)`. The `ConfigureAwait` method is not required for awaitable types, so unless other accommodations are made this feature would not be useful in some scenarios.

**Comment by benaadams** [comment]
> The ConfigureAwait method is not required for awaitable types, so unless...

:sadpanda: I guess it would have to be this then; in which case is only one syntax

```csharp
var addresses = context.GetOrderAsync(id: 5).ConfigureAwait(false)
                      ->CustomerAsync().ConfigureAwait(false)
                      ->AddressesAsync().ConfigureAwait(false)
                      ->FirstOrDefault();
```

**Comment by sharwell** [comment]
@benaadams You can also take the approach of setting the current synchronization context to null, or use a synchronization context which always executes on a particular task scheduler.

**Comment by YaakovDavis** [comment]
What about the following syntax?  

```
var addresses = await context.GetOrderAsync(id: 5)
                then @.CustomerAsync()
                then @.AddressesAsync().FirstOrDefault();
```

Note: `then` clauses can be used even without an implicit context placeholder:

```
var x  =  await Foo() then Bar(); 
```

The above is equivalent to:
```
await Foo();
var x = await Bar();
```

**Comment by dadhi** [comment]
```csharp
var firstAddress = (await GetCustomerAsync(id))
    .Do(async x => (await GetAddressesAsync(x))
    .Do(x => x.FirstOrDefault());
```

Where `Do` is just an extension method with either `Func<T, Task<R>>` or `Func<T, R>` as described here: https://github.com/dotnet/csharplang/issues/74#issuecomment-390212381

**Comment by yaakov-h** [comment]
@dadhi plus overloads for `ValueTask` and any other custom awaitables?

**Comment by dadhi** [comment]
@yaakov-h yep. whatever you need. 

I am using such extensions in couple of code-bases already, including big enterprise production. Given it such a small thing - it may simplify a lot. Especially together with LINQ methods. Often producing clean expression-bodied one-line methods at the end.

Considering performance affected by lambda creation and usage, some of it may be mitigated by passing method group directly or using local method. I think inclusion of static / lightweight delegates in language should help a lot in this and other scenarios.

**Comment by canton7** [comment]
See also the discussion in #35

**Comment by Timo-Weike** [comment]
What is the state of this proposal any chance of it being implemente in the next C# version?

**Comment by jzabroski** [comment]
Bold opinion: I think that the Task API as it exists today is likely not the Task API of the future, so I would caution against tightly coupling syntactic sugar to it.  As an example of a more powerful task API, see Scala's zio, which guarantees resource finalization at the moment the task is complete or canceled or thrown from, and provides the "chain of reasoning" for what task should have "continued with next" had the task not been canceled or thrown from.

The current Task API has a lot of FAQs about how to use it correctly, which mostly boil down to "async all the way" as the best practice.

Keep in mind, from a historical perspective, in 2007, researchers were beginning to socialize the concept that a confluence of Power wall, ILP wall and slow clock frequency growth would combine to force new programming paradigms.

> “You can imagine a scenario where people stop buying laptops and PCs because we can’t figure this out.” 
> 
> David Patterson, 2007, computer-architecture expert and computer science professor at the University of California, Berkeley.

Even in early 2006, UC Berkeley computer scientists wrote a spate of papers articulating various solutions to the "coming crisis", including "The Problem with Threads".  Charles Leierson and his group at Intel had developed the concept of work + span for modeling the cost of Amdahl's Law when decomposing computations represented as directed acyclic graphs into work-stealing dequeues.  And the Task API emerged around that time, with the Parallel.ForEach and until George Chrysanthakopoulos work at Microsoft in 2008 on the Concurrency and Coordination Runtime, MSBuild was basically a single-threaded build engine.  So, we're at best about 14 years into the development of "Task APIs", but really 2008 was the watershed year where Microsoft kicked us in the butt and said, "START WRITING MULTICORE PROGRAMS".  I would guess that the next 14 years yield another two function steps improvement by focusing on debuggability and resource reasoning.  The JVM world has already started to move on by some adapting CatsEffects and zio to model computations and solve these problems.  .NET is a function step behind and still has this dreaded SynchronizationContext ambient value that changes depending on where you call your code from, rather than letting the caller say this is a pure unit of io, but the input comes from this impure source, and the output goes to this impure source.

**Comment by dadhi** [comment]
@Timo-Weike Meanwhile in a [macros world](https://github.com/qwertie/ecsharp) it is a one-liner

```cs
define operator.($x, awaited) => (await($x));
```

Here is the [discussion](https://github.com/qwertie/ecsharp/issues/113#issuecomment-657097287).

**Comment by CyrusNajmabadi** [comment]
That's a good example of why I would generally be against macros in the language.
